enhancements/security/runtime-class-scc.md

```diff
@@ -0,0 +1,248 @@
+# openshift scc runtime class enhancement
+---
+title: gate-runtime-classes-with-scc
+authors:
+  - "@haircommander"
+  - "@mrunalp"
+reviewers:
+  - "@deads2k"
+  - "@sttts"
+approvers:
+  - ""
+creation-date: 2020/12/04
+last-updated: 2020/12/11
+status: provisional
+
+---
+# Gate Runtime Classes with SCC
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Enhance the Openshift SCC with two fields called `AllowRuntimeClasses` and `EnsureRuntimeClasses`, which gate users from specifying runtime classes in their Pod requests. 
+
+## Motivation
+
+Enhancing SCC to be runtime class aware allows admins to customize how their users are able to access different runtime customizations. This is key in enabling new features like unprivileged builds, kata containers, and performance enhancements for latency sensitive workloads, while also providing protection from less privileged users gaining access to features they should not.
+
+### Goals
+
+- SCC should support specifying the runtime classes that a user is allowed access to in their pod specs.
+- SCC should support specifying the runtime classes that must be specified in the pod request from a user.
+
+### Non-Goals
+
+- Changing the current SCC profiles to allow any extra runtime classes.
+- Changing behavior of the default runtime class.
+- Adding additional SCCs for various customzied behavior using AllowedRuntimeClasses.
+
+## Proposal
```

Comment:
you need to describe how the ordering of SCCs will be done.

Reply:
I'm sorry, I don't really understand what you mean by this. Do you have a pointer to a similar description?

Reply:
https://github.com/openshift/apiserver-library-go/blob/master/pkg/securitycontextconstraints/util/sort/bypriority.go is the order @deads2k is talking about, defined through points https://github.com/openshift/apiserver-library-go/blob/6f1013f42f98d74a6b368d670215493f8425194f/pkg/securitycontextconstraints/util/sort/byrestrictions.go#L107.

Reply:
Probably you want some subset order. If one is not the subset of the other, maybe set size matters, and if that is equal you could take something arbitrary but deterministic like alphabetically of the alphabetically first element.

Reply:
Okay I have added a section and some ideas. A subset order makes sense, possibly weighing each value similar to how RunAsUser values are weighted https://github.com/openshift/apiserver-library-go/blob/6f1013f42f98d74a6b368d670215493f8425194f/pkg/securitycontextconstraints/util/sort/byrestrictions.go#L129..L134